Only allow a single pending channel per-peer #32

AndrewSamokhvalov opened the pull request:
**Issue:** [Only allow a single pending channel per-peer #32] (https://github.com/lightningnetwork/lnd/issues/32)

Roasbeef commented on fundingmanager.go:
```diff
@@ -681,6 +705,10 @@ func (f *fundingManager) initFundingWorkflow(targetPeer *peer, req *openChanReq)
 // wallet, then sends a funding request to the remote peer kicking off the
 // funding workflow.
 func (f *fundingManager) handleInitFundingMsg(msg *initFundingMsg) {
+	if !f.checkMaxPending(msg.peer) {
```
In this case, since a local RPC client initiated the request, an error should be returned via the provided error channel within the message. 

So: 
```
msg.err <- fmt.Errorf("some error you think of")
return
```

Roasbeef commented on fundingmanager.go:
```diff
@@ -176,6 +177,25 @@ type numPendingReq struct {
+// is violated.
+func (f *fundingManager) checkMaxPending(peer *peer) bool {
+	if len(f.activeReservations[peer.id]) >= maxActiveReservations {
+		errMsg := &lnwire.ErrorGeneric{
```
This error shouldn't be unconditionally sent since in the case of a client using the `OpenChannel` RPC, the remote peer hasn't yet received any funding negotiation messages.

Roasbeef commented on fundingmanager.go:
```diff
@@ -176,6 +177,25 @@ type numPendingReq struct {
+			ErrorID: lnwire.ErrorMaxPendingChannels,
+		}
+
+		peer.queueMsg(errMsg, nil)
```
The observance of error messages hasn't yet been fleshed out within the `peer` code. Depending on the severity of the error, if the error specifically effects an open channel, then the `channelManager` should initiate a unilateral closure of the channel by broadcasting the latest state. However, in many cases like this one, no such action needs to be carried out, only reporting the error is sufficient.

Roasbeef commented on fundingmanager.go:
```diff
@@ -176,6 +177,25 @@ type numPendingReq struct {
+// checkMaxPending check number of pending channels to not exceed maximum value
+// send the lnwire.ErrorGeneric message to the offending peer if the constraint
+// is violated.
+func (f *fundingManager) checkMaxPending(peer *peer) bool {
```
I think this function should have its scope restricted a bit. Rather than sending an error to the targeted peer if the number of pending channels exceeds the limit, only the bool value should be returned without any further action. 

So the function can mostly be reduced to (depending on the name of the method): 
```go
return len(f.activeReservations[peer.id]) >= maxActiveReservations
```

Roasbeef commented on fundingmanager_test.go:
```diff
@@ -0,0 +1,43 @@
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+func TestMaxPendingPeerExceed(t *testing.T) {
```
In response to the above comments, this test should be changed accordingly. I'd prefer the addition of a test-case using the integration testing framework. 

If you don't think it's too out of scope for this PR, I'd also like to see the `ErrorGeneric` message extended to add a `PendingChannelID` field`. The addition of such a field will let peers communicate errors in the context of a particular pending channel. With this field, once a peer reads an `ErrorGeneric` message with the `PendingChannelID` field set, then they can forward the error to the `fundingManager` who can handle it properly.

Roasbeef commented on lnwire/single_funding_open_proof.go:
```diff
@@ -9,7 +9,8 @@ import (
 // a proof of an open channel to the responder. Otherwise, responding node may
-// be vulernable to a resource exhasution attack wherein the a requesting node
+// be vulnerable to a resource exhaustion attack wherein the a requesting
+// node
```
Can you fix the character folding below this? Within the codebase comments are typically limited to 80 characters.

Thanks!

AndrewSamokhvalov commented on lnd_test.go:
```diff
@@ -579,12 +590,109 @@ func testMultiHopPayments(net *networkHarness, t *testing.T) {
 	"channel balance":        testChannelBalance,
 	"single hop invoice":     testSingleHopInvoice,
-	"test mult-hop payments": testMultiHopPayments,
+	//"test mult-hop payments": testMultiHopPayments,
```
Temporary comment this test almost always fails (in master branch), increasing the timeout not working : (

AndrewSamokhvalov commented:
This PR can't be merged right now because:
1. I commented `testMultiHopPayments` test, because it almost always fail :(
2. In `testMaxPendingChannels` test  I check only local response from `lnd`, and for now I am thinking about how should I check the remote response. The problem is when I create the max number of pending channels between `Alice` and `Bob` I can't send funding request because local check of max pending channels prevents me to do this.
3. There is no logic for handling `lnwire.ErrorGeneric`.

AndrewSamokhvalov commented:
* Added `--pendingchannels` flag in order to have ability control max pending channels number during integration tests. 
* Added transmission of`lnd`process errors in main process.
* Rewrite tests to have one point for error handling (I guess It might worth it, in case we would consider make tests parallel one day, and also it might add readability, but of course the last word for you).

Can the processing of `lnwire.ErrorGeneric` be segregated in separate task? Just to have feeling of milestone :)

Roasbeef commented on fundingmanager.go:
```diff
@@ -684,6 +719,12 @@ func (f *fundingManager) initFundingWorkflow(targetPeer *peer, req *openChanReq)
 // wallet, then sends a funding request to the remote peer kicking off the
 // funding workflow.
 func (f *fundingManager) handleInitFundingMsg(msg *initFundingMsg) {
+	if !f.checkMaxPending(msg.peer) {
```
I think the check for max pending channels here should be removed. 

As suggested above, this option should only constrain the maximum number of _incoming_ pending channels, instead of cumulatively across both outgoing and incoming channels. 

Since this parameter is node policy, different nodes on the network may have a different set policy number. Therefore, as this policy isn't currently signaled anywhere, we must relay on a negative error message from the connected peer in order to properly limit the number of pending outgoing channels we make to that node.

Roasbeef commented on fundingmanager.go:
```diff
@@ -754,3 +796,37 @@ func (f *fundingManager) handleInitFundingMsg(msg *initFundingMsg) {
+		chanID := fmsg.err.PendingChannelID
+
+		f.resMtx.RLock()
+		resCtx := f.activeReservations[peerID][chanID]
```
A two-value read (`resCtx, ok`) should be performed here as the remote peer might send an error message for a pending channel we don't currently know of. In the case of (`!ok`) the instance should be logged, then returned early in order to avoid causing a panic below.

Roasbeef commented on fundingmanager.go:
```diff
@@ -754,3 +796,37 @@ func (f *fundingManager) handleInitFundingMsg(msg *initFundingMsg) {
+		resCtx := f.activeReservations[peerID][chanID]
+		f.resMtx.RUnlock()
+
+		if err := resCtx.reservation.Cancel(); err != nil {
```
In addition to canceling the reservation, the `resCtx` should also be removed from the `activeReservations` map.

Roasbeef commented on lnd.go:
```diff
@@ -135,10 +135,20 @@ func lndMain() error {
 	signer := wc
 	bio := wc
 
+	lightningWalletConfig := &lnwallet.Config{
```
As mentioned above, I see no reason as to why this new configuration parameter would need to be passed into the `lnwallet`.

Roasbeef commented on lnwallet/config.go:
```diff
@@ -12,4 +17,19 @@ type Config struct {
+	// peer. In both cases - locally (RPC client) and remotely
+	// (Node interaction), error will be returned if the number of pending
+	// channels will exceed maximum.
+	MaxPendingChannels int
```
Since this a p2p policy configuration parameter, the wallet doesn't need to know of it. Therefore, the changes in this file should be reverted.

Roasbeef commented on lnwallet/wallet.go:
```diff
@@ -233,13 +233,16 @@ type channelOpenMsg struct {
 // at all, as SPV support is integrated inot btcwallet.
 type LightningWallet struct {
+
+	Cfg           *Config
```
The changes in this file, other than the typo fixes should be reverted as the config changes aren't needed.

Roasbeef commented on lnd_test.go:
```diff
@@ -16,109 +13,181 @@ import (
+// - have ability to properly wrap errors in order to see stack trace.
+// - have nice and elegant way to handle lnd process errors in one
+// select structure with test cases.
+type CT struct {
```
Very nice! I really like the addition of this wrapper struct. Currently, when I run into daemon related errors locally, I make some local changes to `networkHarness` to allow me to peer into the logs and debug the issue manually. With the addition of this wrapper struct, errors propagate cleanly, providing a better user experience for developers writing new integration tests.

Roasbeef commented on lnd_test.go:
```diff
@@ -16,109 +13,181 @@ import (
+		ct.errChan <- errors.Wrap(err, 1).ErrorStack()
+		ct.FailNow()
+	} else {
+		ct.Fatal("can't sen error when test isn't running")
```
Will this error case ever be hit in practice? From a cursory glance, it seems that either of the error methods (this one, and the one below) will never be called outside of test execution.

AndrewSamokhvalov replied:
I added this check in order to exclude the cases when developer might send error in `nil` channel (channel becomes not nil only during test case execution).

Roasbeef commented on lnd_test.go:
```diff
@@ -574,88 +655,223 @@ func testMultiHopPayments(net *networkHarness, t *testing.T) {
+		select {
+		// Receive both types of err - panic and fatal from
+		// one channel and raise the fatal in main "thread".
+		case err := <- errChan:
```
A two-value read can be done here (`err, ok`) in order to check if the channel was closed, or an error was propagated up. So when `!ok` you know the test succeeded, and can move forwards.

Roasbeef commented on lnd_test.go:
```diff
@@ -574,88 +655,223 @@ func testMultiHopPayments(net *networkHarness, t *testing.T) {
 
-		currentTest = testName
-		lnTest(lightningNetwork, t)
+out:
```
Why is this label necessary? It seems on test failure, that the function can simply return as there's no further logic logic after this for-loop.

Roasbeef commented on lnd_test.go:
```diff
@@ -16,109 +13,181 @@ import (
+
+	// Channel for sending retransmitted panic errors and fatal error which
+	// happens in test case.
+	errChan chan interface{}
```
Why is this a `chan interface{}`? More concretely, it seems that a `chan error` is needed instead since only errors will be passed across this channel.

Roasbeef commented on lnd_test.go:
```diff
@@ -574,88 +655,223 @@ func testMultiHopPayments(net *networkHarness, t *testing.T) {
+
+// testMaxPendingChannels checks that in case excess max pending channels for
+// one peer error is returned to the RPC client (in case of
+// local open channel initiation) and to the remote peer (in case of remote
```
As mentioned above, the scope of this test can be reduced to only test against the case a remote (inbound) max pending channel violation.

Roasbeef commented on lnwallet/interface_test.go:
```diff
@@ -321,8 +321,10 @@ func createTestWallet(tempTestDir string, miningNode *rpctest.Harness,
 
-	wallet, err := lnwallet.NewLightningWallet(cdb, notifier, wc, signer,
-		bio, netParams)
+	cfg := &lnwallet.Config{}
```
Similarly, the changes to this file can be reverted as the wallet doesn't need to know of this new config paramter.

Roasbeef commented on networktest.go:
```diff
@@ -159,14 +161,27 @@ func (l *lightningNode) genArgs() []string {
 		return err
 	}
 
+	go func() {
```
Excellent addition! This should make debugging errors when writing integration tests a much, much more pleasant experience.

Roasbeef commented on networktest.go:
```diff
@@ -234,7 +249,14 @@ func (l *lightningNode) cleanup() error {
+	// - start of the node wasn't initiated
+	// - process wasn't spawned
+	// - process already finished
+	if 	l.cmd == nil ||
```
I still need to get up a document detailing contribution guidelines, but this file should be run through `go-fmt`.

AndrewSamokhvalov commented on config.go:
```diff
@@ -56,6 +57,8 @@ type config struct {
 
 	PeerPort int    `long:"peerport" description:"The port to listen on for incoming p2p connections"`
+
+	// TODO(andrew.shvv) Add RPCHost?
```
Sorry, I think I should be more verbose with this `TODO`:

That is true - we have `RPCHost`, `RPCPort` variables, but such naming introduce misconception - `RPCPort` is used for `lnd` initialisation, and for now we haven't way for changing the `lnd` `RPC` host.
```
// Next, Start the grpc server listening for HTTP/2 connections.
grpcEndpoint := fmt.Sprintf("localhost:%d", loadedConfig.RPCPort)
```

and `RPCHost` is used for `btcd` intialiasation:
```
btcdHost := fmt.Sprintf("%v:%v", cfg.RPCHost, activeNetParams.rpcPort)
```

So I think it would be better at least rename parameters.

AndrewSamokhvalov commented on fundingmanager.go:
```diff
@@ -684,6 +719,12 @@ func (f *fundingManager) initFundingWorkflow(targetPeer *peer, req *openChanReq)
 // funding workflow.
 func (f *fundingManager) handleInitFundingMsg(msg *initFundingMsg) {
+	if !f.checkMaxPending(msg.peer) {
+		msg.err <- grpc.Errorf(OpenChannelFundingError, "Number of " +
```
In case of using `grpc.Errorf` we may introduced new error code `OpenChannelFundingError` which helps to determine whether we receive exactly what expected to or not. I think it might be quite useful in near future when will want to distinguish different type of error on `RPC` client side.

```
	_, err = net.OpenChannel(ctx, net.Alice, carol, amount, 1)
	if err == nil {
		ct.Errorf("error wasn't received")
	} else if grpc.Code(err) != OpenChannelFundingError {
		ct.Errorf("not expected error was received : %v", err)
	}
```

AndrewSamokhvalov commented:
Sorry, looks like I need more practise with `rebase`.

Roasbeef commented on fundingmanager.go:
```diff
@@ -754,3 +785,46 @@ func (f *fundingManager) handleInitFundingMsg(msg *initFundingMsg) {
+					"pending channels exceed maximum")
+		}
+
+		f.resMtx.RLock()
```
Since you're performing a delete from the map here (a write), you need to grab the full lock (`.Lock()`) instead of only the read lock.

AndrewSamokhvalov replied:
thanks!
